blockio-api/src/System/FS/BlockIO/API.hs

```diff
@@ -170,3 +210,40 @@ hAdviseAll hbio h advice = hAdvise hbio h 0 0 advice -- len=0 implies until the
 -- present.
 hDropCacheAll :: HasBlockIO m h -> Handle h -> m ()
 hDropCacheAll hbio h = hAdviseAll hbio h AdviceDontNeed
+
+{-------------------------------------------------------------------------------
+  File locks
+-------------------------------------------------------------------------------}
+
+-- | A handle to a file locked using 'tryLockFile'.
+newtype LockFileHandle m = LockFileHandle {
+    -- | Release a file lock acquired using 'tryLockFile'.
+    hUnlock :: m ()
+  }
+
+tryLockFileIO :: HasFS IO HandleIO -> FsPath -> GHC.LockMode -> IO (Maybe (LockFileHandle IO))
+tryLockFileIO hfs fsp mode = do
+    fp <- FS.unsafeToFilePath hfs fsp -- shouldn't fail
+    rethrowFsErrorIO hfs fsp $
+      bracketOnError (GHC.openFile fp GHC.WriteMode) GHC.hClose $ \h -> do
+        bracketOnError (GHC.hTryLock h mode) (\_ -> GHC.hUnlock h) $ \b -> do
```

Comment:
Do we need the second `bracketOnError` with `hUnlock` or is it enough to rely on `hClose`, which should also release the lock? Same for the unlock+close below.

We can just leave it as is, though. It's probably clearer.

Reply:
I think closing the handle *should* release the lock, but alas I'm not sure. It probably doesn't hurt to be extra safe by doing `bracketOnError` twice